src/CarbonAware.DataSources/CarbonAware.DataSources.ElectricityMaps/test/Client/ElectricityMapClientTests.cs

```diff
@@ -0,0 +1,457 @@
+            .BuildServiceProvider();
+        var client = serviceProvider.GetRequiredService<IElectricityMapsClient>();
+
+        // Act & Assert
+        Assert.ThrowsAsync<HttpRequestException>(async () => await client.GetCurrentForecastAsync("lat", "long"));
+    }
+
+    [Test]
+    public void ClientProxyTest_With_Missing_ProxyURL_ThrowsException()
+    {
+        // Arrange
+        var settings = new Dictionary<string, string> {
+            { ForecastDataSourceKey, ForecastDataSourceValue },
+            { HeaderKey, AuthHeader },
+            { TokenKey, DefaultTokenValue },
+            { UseProxyKey, "true" },
+        };
+        var configuration = new ConfigurationBuilder()
+            .AddInMemoryCollection(settings)
+            .AddEnvironmentVariables()
+            .Build();
+        var serviceCollection = new ServiceCollection();
+
+        // Act & Assert
+        Assert.Throws<ConfigurationException>(() => serviceCollection.AddElectricityMapsForecastDataSource(configuration.DataSources()));
+    }
+
+    private void CreateBasicClient(string zoneContent, string resultContent)
+    {
+        this.CreateHttpClient(m =>
+        {
+            var isTokenAuthValid = m.RequestUri!.ToString().Contains(BaseUrls.TokenBaseUrl) && !m.Headers.Where(x => x.Key == "auth-token").Any();
+            var isTrialAuthValid = m.RequestUri!.ToString().Contains(BaseUrls.TrialBaseUrl) && !m.Headers.Where(x => x.Key == "X-BLOBR-KEY").Any();
+            // If no auth and token setup, return unauthorized
+            if (isTokenAuthValid || isTrialAuthValid)
+            {
+                return Task.FromResult(new HttpResponseMessage(System.Net.HttpStatusCode.Unauthorized));
+            }
+
+            if (m.RequestUri?.ToString() == BaseUrls.TokenBaseUrl + Paths.Zones)
+            {
+                var response = this.MockElectricityMapsResponse(m, new StringContent(zoneContent));
+                return Task.FromResult(response);
+            }
+            return Task.FromResult(this.MockElectricityMapsResponse(m, new StringContent(resultContent)));
+        });
+    }
+
+    private HttpResponseMessage MockElectricityMapsResponse(HttpRequestMessage request, HttpContent responseContent)
+    {
+        var response = new HttpResponseMessage(System.Net.HttpStatusCode.OK);
+        response.Content = responseContent;
+        return response;
+    }
+
+    private void CreateHttpClient(Func<HttpRequestMessage, Task<HttpResponseMessage>> requestDelegate)
+    {
+        this.MessageHandler = new MockHttpMessageHandler(requestDelegate);
+        this.HttpClient = new HttpClient(this.MessageHandler);
+        this.HttpClientFactory = Mock.Of<IHttpClientFactory>();
+        Mock.Get(this.HttpClientFactory).Setup(h => h.CreateClient(IElectricityMapsClient.NamedClient)).Returns(this.HttpClient);
+    }
+
+    private class MockHttpMessageHandler : HttpMessageHandler
```

Comment:
Consider using `Moq.Protected` to manage the setup for `SendAsync` rather than writing our own class

- [Moq/moq4#wiki](https://github.com/Moq/moq4/wiki/Quickstart#miscellaneous:~:text=Moq%204.8%20and%20later%20allows%20you%20to%20set%20up%20protected%20members)
- https://kagamine.dev/en/mock-httpclient-the-easy-way/

Reply:
This is to mirror the WattTimeClient, would it be worth doing that as a separate cleanup on both later?

Reply:
@bderusha  will circle back

Reply:
Noted elsewhere for future cleanup.